api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto

```diff
@@ -22,4 +24,9 @@ message PostgresProxy {
   // The human readable prefix to use when emitting :ref:`statistics
   // <config_network_filters_postgres_proxy_stats>`.
   string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
+
+  // Controls whether SQL statements received in Frontend Query messages
+  // are parsed. Parsing is required to produce Postgres proxy filter
+  // metadata. Defaults to true.
+  google.protobuf.BoolValue enable_sql_parsing = 2;
```

Comment:
Presumably this doesn't change the wire behavior today? I.e. existing users won't be caught off-guard. Is the performance different significant?

Reply:
Parsing incoming SQL statements is a new feature so there is no problem with existing users.

There is long story why this flag is introduced and it was discussed on envoy-postgres slack channel with @fabriziomello and @ahachete. Performance was a major driver behind that. The other issue is that the current SQL parser which we inherited from MySQL parses successfully only simple queries but is not able to parse vast number of more complex Postgres SQL statements. The plan is to link to Postgres source code for parsing SQL statements, but this will not happen immediately.

Reply:
@htuch I'll run some extensive tests using Postgres regression suite against Envoy and get some numbers turning `on` and `off` the parsing. But as @cpakulski mentioned isn't the performance the only reason to introduce this flag. I ran some tests today and ~70% of the Postgres queries are not parsed so we'll need to improve it in the future.

Reply:
Yeah, I don't to create a ton of work here, my main observation was that this is *default on*, which would impact existing users if there was a performance or reliability degradation. Given most queries aren't parsed, should it rather be *default off*?

Reply:
Postgres is experimental filter still under active development. Envoy release 1.15 will be the very first release including this filter, so there should not be any issues from existing users.
All new features planned for this Postgres filter (measuring latency per query type, routing based on query, enforcing RBAC policies) require SQL parsing to be enabled. There will be some scenarios where a user may wish to disable SQL parsing (like using the filter only to terminate SSL), but filter will work even if SQL is enabled, albeit with some performance cost. 
I agree that at the moment most SQL parsing fails, but what passes is enough for experimentation and demonstration.
In the long run I believe that most applications of this filter will require that flag to be true (that is why the default option is TRUE). If we go ahead with default FALSE, we may need breaking API change in the future.

Reply:
@cpakulski this PR introduce two new counters `queries_parsed` and `queries_parse_error` but we already have `statements_*` counters, so IMHO it should be renamed to `statements_parsed` and `statements_parse_error`.

Reply:
Ack, API looks good then.

Reply:
what is the purpose of having a "decode_this_protocol" in a protocol decoder for that protocol? It seems redundant isn't it? When we did the mysql one, it was always parsing the SQL query. if it couldn't, it falls back to a simple passthrough model. To @htuch's point, shouldn't the default be on with some explicit flag to turn it off if some user desires to do so? Since the entire filter is experimental, there is no breaking change here either.

Reply:
We currently parse Backend messages (messages returned by the server to a client). They carry confirmation that specific operations has been successfully completed (SELECT, COMMIT, ROLLBACK, INSERT, UPDATE). 
Parsing SQL queries is not required to produce those stats. 

Parsing SQL query is needed for the following scenarios:
- produce metadata (this PR addresses this)
- route based on query
- latency per query type
- errors per query type

The problem is that current SQL parser does not parse extended and prepared Postgres messages and ~40% of Basic query messages. It is enough for simple demonstration and proof--of-concept, but is not sufficient for production. We hope that we can use Postgres open source code for parsing SQL queries and link it to Envoy.

After discussions here and on slack it all comes to one thing: whether to make parsing SQL on or off by default.
In the long run I believe that most features in this filter will require it to be on, but in the short term maybe it makes sense to make it off by default.
Since it is experimental filter, can we change API in the future and change the default value from off to on?

Reply:
The point here is consistency with the mysql and mongo API. Producing metadata is still a best-effort read operation that should be treated in same vein as stats. If it requires SQL parsing, then SQL parsing should be on by default. What you need a feature flag for is the "route based on query". Metrics shouldn't require a feature flag IMO. If we can parse, we emit metrics. If we cannot parse, we `omit` the metric.

If we have the enable-sql parsing on, we should have this for mysql as well